chall04/gbudau.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+
+import sys
+
+STDIN = 0
+FILE = 1
+
+def read_error(filename):
+    print("{}: {}: Can't read file".format(sys.argv[0], filename))
+
+def space_error(filename):
+    print("{}: {}: Not enough space in the given shelves".format(sys.argv[0], filename))
+
+def fill_bookshelves(filename, bookshelves, books):
+    # Convert books and bookshelves to integers
+    # Sort from max to min
+    books = [int(i) for i in books]
+    bookshelves = [int(i) for i in bookshelves]
+    books.sort(reverse = True)
+    bookshelves.sort(reverse = True)
+
+    used_bookshelves = []
+    while books and bookshelves and ((bookshelves[0] >= books[0]) or (used_bookshelves and used_bookshelves[0] >= books[0])):
```

Comment:
I got this a bit mess up, should be:
`while books and ((bookshelves and bookshelves[0] >= books[0]) or (used_bookshelves and used_bookshelves[0] >= books[0])):`

Reply:
It breaks in cases where the numbers of bookshelves is less than the number of books.
```
1000
70 A Game of Thrones
76 A Clash of Kings
99 A Storm of Swords
75 A Feasts for Crows
```
Without parenthesis it will check if there are any `bookshelves` left but in this cases they have been moved to `used_bookshelves`
With parenthesis checks both `bookshelves` and `used_bookshelves`.